prov/util/src/util_profile.c

```diff
@@ -0,0 +1,304 @@
+ * SOFTWARE.
+ *
+ */
+
+#include <stdarg.h>
+#include <stdio.h>
+#include <stdlib.h>
+
+#include <ofi.h>
+#include <ofi_profile.h>
+
+#define PROF_LIST_SIZE	64
+
+#define OFI_PROF_DESC_SET(dest, src)                 \
+	do {                                             \
+		(dest)->id = (src)->id;                      \
+		(dest)->datatype_sel = (src)->datatype_sel;  \
+		(dest)->datatype = (src)->datatype;          \
+		(dest)->flags = (src)->flags;                \
+		(dest)->size = (src)->size;                  \
+		(dest)->name = (src)->name;                  \
+		(dest)->desc = (src)->desc;                  \
+	} while (0)
+
+struct fi_profile_desc  ofi_common_vars[] = {
+	{.id = FI_VAR_UNEXP_MSG_CNT,
+	 .datatype_sel = fi_primitive_type,
+	 .datatype.primitive = FI_UINT64,
+	 .size = 8,
+	 .name = "pvar_unexp_msg_cnt",
+	 .desc = "Unexpected Receive Message Count"
+	},
+	{.id = FI_VAR_UNEXP_MSG_QUEUE,
+	 .datatype_sel = fi_defined_type,
+	 .datatype.defined = FI_TYPE_CQ_ERR_ENTRY,
+	 .size = sizeof(struct fi_cq_err_entry),
+	 .name = "pvar_unexp_msg_queue",
+	 .desc = "Unexpected Receive Message Queue"
+	},
+};
+
+struct fi_profile_desc  ofi_common_events[] = {
+	{.id = FI_EVENT_UNEXP_MSG_RECVD,
+	 .datatype_sel = fi_defined_type,
+	 .datatype.defined = FI_TYPE_CQ_ERR_ENTRY,
+	 .size = sizeof(struct fi_cq_err_entry),
+	 .name = "pevent_unexp_msg_recd",
+	 .desc = "Unexpected Message Received"
+	},
+	{.id = FI_EVENT_UNEXP_MSG_MATCHED,
+	 .datatype_sel = fi_defined_type,
+	 .datatype.defined = FI_TYPE_CQ_ERR_ENTRY,
+	 .size = sizeof(struct fi_cq_err_entry),
+	 .name = "pevent_unexp_msg_matched",
+	 .desc = "Unexpected Message Matched"
+	},
+};
+
+size_t ofi_common_var_count = ARRAY_SIZE(ofi_common_vars);
+size_t ofi_common_event_count = ARRAY_SIZE(ofi_common_events);
+
+int ofi_prof_pcb_noop(struct fid_profile *prof_fid,
+                      struct fi_profile_desc *event,
+                      void *param, size_t size, void *context)
+{
+	OFI_UNUSED(prof_fid);
+	OFI_UNUSED(event);
+	OFI_UNUSED(param);
+	OFI_UNUSED(size);
+	OFI_UNUSED(context);
+	return -FI_ENOSYS;
+}
+
+int ofi_prof_init(struct util_profile *prof, struct fid *fid,
+                  uint64_t flags, void *context, struct fi_profile_ops *ops,
+                  int prov_vars_size, int prov_events_size)
+{
+	prof->prof_fid.fid.fclass = FI_CLASS_PROFILE;
+	prof->prof_fid.fid.context = context;
+	prof->prof_fid.ops = ops;
+	prof->fid = fid;
+	prof->flags = flags;
+
+	prof->var_count = 0;
+	prof->event_count = 0;
+
+	prof->data_cached = false;
+	prof->varlist_size = ofi_common_var_count + prov_vars_size;
+	prof->vars = (void **)calloc(prof->varlist_size, sizeof(void *));
+	prof->data = (void *)calloc(prof->varlist_size, 
+	                                sizeof(struct util_prof_data));
+	prof->varlist = (struct fi_profile_desc *)calloc(prof->varlist_size,
+	                         sizeof(struct fi_profile_desc));
```

Comment:
Alignment seems to be off. Also make sure to use tab whenever there are 8 consecutive spaces.